src/modules/directory.rs

```diff
@@ -134,6 +136,60 @@ fn contract_path(full_path: &Path, top_level_path: &Path, top_level_replacement:
     )
 }
 
+/// Contract the root component of a path based on the real path
+///
+/// Replaces the `top_level_path` in a given `full_path` with the provided
+/// `top_level_replacement` by walking ancestors and comparing its real path.
+fn contract_repo_path(full_path: &Path, top_level_path: &Path) -> Option<String> {
+    let top_level_real_path = real_path(top_level_path);
+    // Walk ancestors to preserve logical path in `full_path`.
+    // If we'd just `full_real_path.strip_prefix(top_level_real_path)`,
+    // then it wouldn't preserve logical path. It would've returned physical path.
+    for (i, ancestor) in full_path.ancestors().enumerate() {
+        let ancestor_real_path = real_path(ancestor);
+        if ancestor_real_path != top_level_real_path {
+            continue;
+        }
+
+        let components: Vec<_> = full_path.components().collect();
+        let repo_name = components[components.len() - i - 1]
+            .as_os_str()
+            .to_str()
+            .unwrap();
+
+        if i == 0 {
+            return Some(repo_name.to_string());
+        }
+
+        let path = PathBuf::from_iter(&components[components.len() - i..]);
```

Comment:
`PathBuf::from_str` should work here? One less memory allocation.

Reply:
I think there is no way to get sub-`&Path` or `&str` or `&OsStr` from a `full_path.components().skip(N)...`. Similar situation to that `"some string".chars().skip(5)` cannot get &"string"

Reply:
Ah, sorry @foriequal0 - I mistakenly thought that `components` was a string slice.

Your existing implementation is fine 👍